File: scripts/core/pipeline_report.py
#!/usr/bin/env python3
"""Post-run inventory: rulesets, AdBlock shards, modules, SRS coverage."""

from __future__ import annotations

import os
from typing import List, Tuple

from core.common import Logger, get_project_root, read_file

ROOT = get_project_root()
SURGE_DIR = os.path.join(ROOT, "rulesets/surge-shadowrocket")
DNS_MAPPING_DIR = os.path.join(ROOT, "rulesets/Sources/DNS_mapping")
DNS_RULES_DIR = os.path.join(ROOT, "rulesets/dns")
ADBLOCK_DIR = os.path.join(ROOT, "rulesets/AdBlock")
SINGBOX_DIR = os.path.join(ROOT, "rulesets/SingBox")
MODULE_SURGE = os.path.join(ROOT, "modules/surge")
MODULE_SR = os.path.join(ROOT, "modules/shadowrocket")

SRS_SOURCE_DIRS = [SURGE_DIR, DNS_MAPPING_DIR, ADBLOCK_DIR]
SRS_SKIP_PREFIXES = ("AdBlock_",)  # large shards: optional compile, still audited


def _rule_count(path: str) -> int:
    n = 0
    for line in read_file(path):
        s = line.strip()
        if s and not s.startswith("#"):
            n += 1
    return n


def _collect_list_files() -> List[str]:
    files: List[str] = []
    for directory in SRS_SOURCE_DIRS:
        if not os.path.isdir(directory):
            continue
        for name in sorted(os.listdir(directory)):
            if name.endswith(".list"):
                files.append(os.path.join(directory, name))
    return files


def audit_srs_coverage() -> Tuple[int, List[str]]:
    missing: List[str] = []
    total = 0
    for list_path in _collect_list_files():
        base = os.path.splitext(os.path.basename(list_path))[0]
        if any(base.startswith(p) for p in SRS_SKIP_PREFIXES):
            continue
        if _rule_count(list_path) == 0:
            continue
        total += 1
        srs_path = os.path.join(SINGBOX_DIR, f"{base}_Singbox.srs")
        # Some legitimately-small SRS outputs (e.g. ASN) still compile into a
        # valid binary with a stable header "SRS". Avoid false negatives by
        # validating the magic bytes instead of an arbitrary size threshold.
        if not os.path.isfile(srs_path):
            missing.append(f"{base}.list -> {os.path.basename(srs_path)}")
            continue
        try:
            with open(srs_path, "rb") as f:
                magic = f.read(3)
            if magic != b"SRS":
                missing.append(f"{base}.list -> {os.path.basename(srs_path)}")
        except Exception:
            missing.append(f"{base}.list -> {os.path.basename(srs_path)}")
    return total, missing


def print_pipeline_summary() -> None:
    Logger.section("Pipeline Summary")

    surge_lists = [
        f for f in os.listdir(SURGE_DIR) if f.endswith(".list")
    ] if os.path.isdir(SURGE_DIR) else []
    adblock_lists = [
        f for f in os.listdir(ADBLOCK_DIR) if f.endswith(".list")
    ] if os.path.isdir(ADBLOCK_DIR) else []
    srs_files = [
        f for f in os.listdir(SINGBOX_DIR) if f.endswith(".srs")
    ] if os.path.isdir(SINGBOX_DIR) else []

    surge_modules = 0
    for root, _dirs, files in os.walk(MODULE_SURGE):
        surge_modules += sum(1 for f in files if f.endswith(".sgmodule"))

    sr_modules = 0
    if os.path.isdir(MODULE_SR):
        for root, _dirs, files in os.walk(MODULE_SR):
            sr_modules += sum(1 for f in files if f.endswith(".module"))

    Logger.info(f"Surge rulesets: {len(surge_lists)} | AdBlock shards: {len(adblock_lists)}")
    Logger.info(f"Sing-box SRS: {len(srs_files)}")
    Logger.info(f"Surge modules: {surge_modules} | Shadowrocket modules: {sr_modules}")

    checked, missing = audit_srs_coverage()
    if missing:
        Logger.warn(f"SRS coverage: {len(missing)}/{checked} list files missing SRS")
        for item in missing[:15]:
            Logger.warn(f"  Missing: {item}")
        if len(missing) > 15:
            Logger.warn(f"  ... and {len(missing) - 15} more")
    else:
        Logger.success(f"SRS coverage OK ({checked} compiled sources checked)")

File: create/hub/pipeline_report.go
package hub

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

var (
	SRS_SOURCE_DIRS   = []string{RULE_SET_DIR, DNS_MAPPING_DIR, ADBLOCK_DIR}
	SRS_SKIP_PREFIXES = []string{"AdBlock_"}
)

func ruleCount(path string) int {
	n := 0
	lines := ReadFile(path)
	for _, line := range lines {
		s := strings.TrimSpace(line)
		if s != "" && !strings.HasPrefix(s, "#") {
			n++
		}
	}
	return n
}

func collectListFiles() []string {
	var files []string
	for _, directory := range SRS_SOURCE_DIRS {
		if !ValidateDirExists(directory, "") {
			continue
		}
		entries, err := os.ReadDir(directory)
		if err != nil {
			continue
		}
		var names []string
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".list") {
				names = append(names, e.Name())
			}
		}
		for _, name := range names {
			files = append(files, filepath.Join(directory, name))
		}
	}
	return files
}

func AuditSrsCoverage() (int, []string) {
	var missing []string
	total := 0
	listFiles := collectListFiles()

	for _, listPath := range listFiles {
		base := strings.TrimSuffix(filepath.Base(listPath), ".list")
		skip := false
		for _, p := range SRS_SKIP_PREFIXES {
			if strings.HasPrefix(base, p) {
				skip = true
				break
			}
		}
		if skip {
			continue
		}
		if ruleCount(listPath) == 0 {
			continue
		}
		total++

		var srsPath string
		if strings.Contains(listPath, DNS_MAPPING_DIR) || strings.Contains(listPath, "/dns/") {
			srsPath = filepath.Join(SINGBOX_DNS_DIR, fmt.Sprintf("%s_Singbox.srs", base))
		} else if strings.Contains(listPath, ADBLOCK_DIR) {
			srsPath = filepath.Join(ADBLOCK_DIR, fmt.Sprintf("%s_Singbox.srs", base))
		} else {
			srsPath = filepath.Join(SINGBOX_DIR, fmt.Sprintf("%s_Singbox.srs", base))
		}

		if !ValidateFileExists(srsPath, "") {
			missing = append(missing, fmt.Sprintf("%s.list -> %s", base, filepath.Base(srsPath)))
			continue
		}

		f, err := os.Open(srsPath)
		if err != nil {
			missing = append(missing, fmt.Sprintf("%s.list -> %s", base, filepath.Base(srsPath)))
			continue
		}

		magic := make([]byte, 3)
		n, err := f.Read(magic)
		f.Close()

		if err != nil || n < 3 || string(magic) != "SRS" {
			missing = append(missing, fmt.Sprintf("%s.list -> %s", base, filepath.Base(srsPath)))
		}
	}
	return total, missing
}

func PrintPipelineSummary() {
	Section("Pipeline Summary")

	surgeLists := 0
	if ValidateDirExists(RULE_SET_DIR, "") {
		entries, _ := os.ReadDir(RULE_SET_DIR)
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".list") {
				surgeLists++
			}
		}
	}

	adblockLists := 0
	if ValidateDirExists(ADBLOCK_DIR, "") {
		entries, _ := os.ReadDir(ADBLOCK_DIR)
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".list") {
				adblockLists++
			}
		}
	}

	srsFiles := 0
	if ValidateDirExists(SINGBOX_DIR, "") {
		entries, _ := os.ReadDir(SINGBOX_DIR)
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".srs") {
				srsFiles++
			}
		}
	}

	surgeModules := 0
	filepath.Walk(SURGE_MODULE_DIR, func(path string, info os.FileInfo, err error) error {
		if err == nil && !info.IsDir() && strings.HasSuffix(info.Name(), ".sgmodule") {
			surgeModules++
		}
		return nil
	})

	srModules := 0
	if ValidateDirExists(SHADOWROCKET_MODULE_DIR, "") {
		filepath.Walk(SHADOWROCKET_MODULE_DIR, func(path string, info os.FileInfo, err error) error {
			if err == nil && !info.IsDir() && strings.HasSuffix(info.Name(), ".module") {
				srModules++
			}
			return nil
		})
	}

	Info(fmt.Sprintf("Surge rulesets: %d | AdBlock shards: %d", surgeLists, adblockLists))
	Info(fmt.Sprintf("Sing-box SRS: %d", srsFiles))
	Info(fmt.Sprintf("Surge modules: %d | Shadowrocket modules: %d", surgeModules, srModules))

	checked, missing := AuditSrsCoverage()
	if len(missing) > 0 {
		Warn(fmt.Sprintf("SRS coverage: %d/%d list files missing SRS", len(missing), checked))
		maxMissing := len(missing)
		if maxMissing > 15 {
			maxMissing = 15
		}
		for _, item := range missing[:maxMissing] {
			Warn(fmt.Sprintf("  Missing: %s", item))
		}
		if len(missing) > 15 {
			Warn(fmt.Sprintf("  ... and %d more", len(missing)-15))
		}
	} else {
		Success(fmt.Sprintf("SRS coverage OK (%d compiled sources checked)", checked))
	}
}
